File: 2025/4/forklift.rb
#!/usr/bin/env ruby

def count_neighbors_on_line(a, i, j, include_self: false)
  return 0 if i < 0
  return 0 if i >= a.size

  #i %= a.size
  #j %= a[i].size

  n = 0
  n += 1 if j > 0 and a[i][j-1] == '@'
  n += 1 if include_self and a[i][j] == '@'
  n += 1 if j < a[i].size and a[i][j+1] == '@'

  n
end

def count_neighbors(a, i, j)
  n = 0
  n += count_neighbors_on_line(a, i-1, j, include_self: true)
  n += count_neighbors_on_line(a, i, j, include_self: false)
  n += count_neighbors_on_line(a, i+1, j, include_self: true)

  n
end

def doit(a)
  removed = []
  sum = 0
  for i in 0...a.size
    for j in 0...a[i].size
      next if a[i][j] != '@'

      n = count_neighbors(a, i, j)
      p "#{i} #{j} has #{n}"
      if n < 4
        sum += 1
        a[i][j] = 'x'
        #removed.push([i, j])
      end
    end
  end

#  removed.each do |i,j|
#    a[i][j] = 'x'
#  end

  sum
end

def main
  f = open("input.txt")
  sum = 0
  a = []
  f.each_line do |line|
    chars = line.chomp.chars
    a.push(chars)
  end

  while true
    n = doit(a)
    sum += n
    break if n == 0
  end

  p "Sum: #{sum}"
end

main()

File: 2025/4/forklift.rs
fn count_neighbors(a: &Vec<Vec<char>>, i: usize, j: usize) -> i64 {
	let mut n = 0;
	let directions = [
		(-1, -1),
		(-1, 0),
		(-1, 1),
		(0, -1),
		(0, 1),
		(1, -1),
		(1, 0),
		(1, 1)];

	for (di, dj) in directions.iter() {
		let ei = i as isize + di;
		let ej = j as isize + dj;
		if ei >= 0 && ei < a.len() as isize && ej >= 0 && ej < a[ei as usize].len() as isize {
			if a[ei as usize][ej as usize] == '@' {
				n += 1;
			}
		}
	}

	return n;
}

fn doit(a: &mut Vec<Vec<char>>) -> i64 {
	let mut sum = 0;
	for i in 0..a.len() {
		for j in 0..a[i].len() {
			if a[i][j] == '@' {
				let n = count_neighbors(&a, i, j);
				if n < 4 {
					sum += 1;
					a[i][j] = 'x';
				}
			}
		}
	}
	return sum;
}

fn main() {
	use std::fs::File;
	use std::io::{BufRead, BufReader};

	let mut sum: i64 = 0;

	let f = File::open("input.txt").unwrap();
	let reader = BufReader::new(f);

	let mut a: Vec<Vec<char>> = Vec::new();

	for line in reader.lines() {
		let s = line.unwrap();
		a.push(s.chars().collect());
	}

	loop {
		let n = doit(&mut a);
		sum += n;
		if n == 0 {
			break;
		}
	}

	println!("Sum: {}", sum);
}
